index.d.ts

```diff
@@ -1,288 +1,600 @@
+    getToggleAllRowsSelectedProps: (props?: object) => object
+    isAllRowsSelected: boolean
+  }
+
+  export interface UseRowSelectRowProps<D extends object> {
+    isSelected: boolean
+    toggleRowSelected: (set?: boolean) => void
+    getToggleRowSelectedProps: (props?: object) => object
+  }
+  /* #endregion */
+
+  /* #region useRowState */
+  export function useRowState<D extends object = {}>(hooks: Hooks<D>): void
+  export namespace useRowState {
+    export const pluginName = 'useRowState'
   }
 
-  export function useGroupBy<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & { rows: Row<D>[] }
+  export type UseRowStateOptions<D extends object> = Partial<{
+    initialRowStateAccessor: (row: Row<D>) => UseRowStateLocalState<D>
+  }>
 
-  export function usePagination<D = {}>(
-    props: TableOptions<D>
-  ): UsePaginationValues<D>
+  export interface UseRowStateState<D extends object> {
+    rowState: Partial<{
+      cellState: UseRowStateLocalState<D>
+      rowState: UseRowStateLocalState<D>
+    }>
+  }
+
+  export interface UseRowStateInstanceProps<D extends object> {
+    setRowState: (rowPath: string[], updater: UseRowUpdater) => void // Purposely not exposing action
+    setCellState: (
+      rowPath: string[],
+      columnID: IdType<D>,
+      updater: UseRowUpdater
+    ) => void
+  }
 
-  export function useExpanded<D = {}>(
-    props: TableOptions<D>
-  ): TableOptions<D> & {
-    toggleExpandedByPath: () => any
-    expandedDepth: []
+  export interface UseRowStateRowProps<D extends object> {
+    state: UseRowStateLocalState<D>
+    setState: (updater: UseRowUpdater) => void
+  }
+  export interface UseRowStateCellProps<D extends object> {
+    state: UseRowStateLocalState<D>
+    setState: (updater: UseRowUpdater) => void
+  }
+
+  type UseRowUpdater<T = unknown> = T | ((prev: T) => T)
+  type UseRowStateLocalState<D extends object, T = unknown> = Record<
+    IdType<D>,
+    T
+  >
+  /* #endregion */
+
+  /* #region useSortBy */
+  export function useSortBy<D extends object = {}>(hooks: Hooks<D>): void
+  export namespace useSortBy {
+    export const pluginName = 'useSortBy'
+  }
+
+  export type UseSortByOptions<D extends object> = Partial<{
+    manualSorting: boolean
+    disableSorting: boolean
+    disableMultiSort: boolean
+    isMultiSortEvent: (e: MouseEvent) => boolean
+    maxMultiSortColCount: number
+    disableSortRemove: boolean
+    disabledMultiRemove: boolean
+    orderByFn: (
+      rows: Row<D>[],
+      sortFns: SortByFn<D>[],
+      directions: boolean[]
+    ) => Row<D>[] // CHECK
+    sortTypes: Record<string, SortByFn<D>>
+  }>
+
+  export interface UseSortByHooks<D extends object> {
+    getSortByToggleProps: ((
+      column: Column<D>,
+      instance: TableInstance<D>
+    ) => object)[]
+  }
+
+  export interface UseSortByState<D extends object> {
+    sortBy: SortingRule<D>[]
+  }
+
+  export type UseSortByColumnOptions<D extends object> = Partial<{
```

Comment:
I don't know how to extend the Column type by this 🤔

Reply:
So the way to use plugins currently is by using interface merging (aka declaration merging). At the top of this file, you will find all the main types we use, and will have to extend those with each plugins' types. You will end up with something like this. https://github.com/stramel/react-table-example-ts/blob/307d0360d0756afd123ca7c8ae68eedf49ca6769/custom_typings/react-table.d.ts#L6-L64